Patch for Modbus_POE_ETH_Relay_30CH and a lot more

icecoldfire opened the pull request:
I found you package for an 8 channel Waveshare Modbus relay, but it doesn't support the 30 channels variant. So I created these patch with these changes:

- Add support for Home Assistant 2025.4
- Configure the channel count of the device
- Add exception messages handling from the Modbus 
- Set relay interval step size from 5 to 1
- Add Dutch translations
- Improve readme.md
- Added extra logging statements
- General code improvements...

Let me know what you think!

Commit:
Add relay channel count option

Commit:
update readme

Commit:
improve validation

icecoldfire committed:
Update sensor.py

icecoldfire committed:
Update sensor.py

Commit:
Improve channel reading

icecoldfire committed:
Update README.md

icecoldfire committed:
Update hacs.json

Commit:
Bugfixes

Commit:
Add error messages

Commit:
Let user reconfigure integration

icecoldfire committed:
Add tests (#1)

Commit:
Add Build Status badge

icecoldfire committed:
Add ruff and mypy (#2)

Commit:
Add translations

posixx commented:
Could this patch be modified so that when interval is 0, the switch stays on permanently? Right now it isn't possible to just switch on and off manually regardless of interval.

icecoldfire commented:
Hey @posixx 

The `0` value is currently used for the flash command, so that wouldn't be possible. But we can maybe use a negative value, for example `-1` to disable the interval logic and stopwatch. 

We also need to expand the modbus command logic with the `on` and `off` command. 

Another solution is just use the Modbus service like documented here: https://www.waveshare.com/wiki/Modbus_POE_ETH_Relay_30CH#Related_applications

Are you currently an active user of the Waveshare Relay? Are there any other features you have missing?

posixx commented:
Yes i'm currently using your fork with the changes within HA. As i'm going to use the module mainly for irrigation a value higher than approx 54 minutes (32767 is the max flash value according to the docs) is needed so your suggestion would be very welcome; set the slider to -1 = disable flash and use on/off commands. I don't like using multiple solutions simultaneously; i had integrated the relay board using modbus intergation but it's a hassle to make a relay flash.

icecoldfire commented:
Nice to hear, I shall make the necessary changes!

icecoldfire committed:
feat: permanent on/off switch (#3)

- If the interval is set to 0 we can now permanent switch the relay on or off
- Extended the CLI and util with the permanent on/off command
- Added the config option for disabling the timers

icecoldfire commented:
@posixx I have merged the changes.

The interval 0, flash logic wasn't correct so I implemented like you proposed. 

One point of improvement in this code base is the handling of the polling, now we only poll if the switch is enabled and then every second. 

Future work is maybe to poll all relay switch at once, but I don't know what a good interval is, because the default 10 or 30 sec can be to slow.

posixx commented:
@icecoldfire So in order to have the switch permanently turn on the interval should be 0?

icecoldfire commented:
Yes, indeed

When configured with the 0 interval it would keep the state permanent.

icecoldfire committed:
Add HACS validation (#4)

icecoldfire committed:
Potential fix for code scanning alert no. 1: Workflow does not contain permissions (#5)

Commit:
Fix format

posixx commented:
Works perfectly, thx!!

jgray-proaudio commented:
Hello. 

Thank you very much for the great work you have done on this. Worked amazing!! 

I was very happy to find setting the interval to 0 is a fixed state, would it be possible to allow for longer than 600 seconds? I have some timers that I would like to use up to 30min for.

icecoldfire commented:
Hey @jgray-proaudio,

The max interval value is theoretically 0xFFFF or 65535, this value needs to be multiplied with 100ms. So the max value would be around 109 minutes.

It would be a small change to allow larger values.

Do you use the stopwatch functionality?

icecoldfire committed:
Increase max interval value and fix step interval (#6)

* Increase max interval value

* increase interval step

---------

Co-authored-by: Stijn Goethals <stijn.goethals@outlook.be>

icecoldfire commented:
@jgray-proaudio the changes are merged

spuky commented:
@icecoldfire love your fork works out of the box.... subscribing to get notified... when merged...
